### PR TITLE
Enable Sequence Packing and Pipeline Parallel in NeVA

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

Enable Sequence Packing and Pipeline Parallel in NeVA

**Collection**: [Multimodal]

# Changelog 
- Added Sequence Packing dataset preparation script
- Update Neva Model and dataset to support packed dataset, also add args to turn on packed dataset
- Enable Pipeline parallelism

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI
To run Jenkins, a NeMo User with write access must comment `jenkins` on the PR.

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [ ] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: examples/multimodal/multimodal_llm/neva/sequence_packing/preprocess_dataset.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
+# Copyright (c) 2023, NVIDIA CORPORATION.  All rights binserved.
```
Comment (cuichenx) [inline examples/multimodal/multimodal_llm/neva/sequence_packing/preprocess_dataset.py:1]: update year to 2024 for new scripts

File: nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py [hunk 2]
```diff
@@ -678,16 +772,40 @@ def fwd_output_and_loss_func(dataloader_iter, model, checkpoint_activations_all_
                 if self.use_loss_mask:
                     forward_args['loss_mask'] = batch['loss_mask']
                 forward_args['checkpoint_activations_all_layers'] = checkpoint_activations_all_layers
+            else:
+                if 'cu_seqlens' in batch:  # packed sequence from GPTSFTPackedDataset
```
Comment (cuichenx) [inline nemo/collections/multimodal/models/multimodal_llm/neva/neva_model.py:776]: update comment here

File: examples/multimodal/multimodal_llm/neva/sequence_packing/preprocess_dataset.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+    """
+    Assigns sequences to bins with shuffling, trying to maximize the packing efficiency.
+
+    Parameters:
+    - seq_lens: List of sequence lengths.
+    - max_seq_length: Maximum capacity of each bin.
+
+    Returns:
+    - List of bins with assigned sequence lengths.
+    """
```
Comment (cuichenx) [inline examples/multimodal/multimodal_llm/neva/sequence_packing/preprocess_dataset.py:190]: can you update the doc string here, i couldn't understand this the first time.. maybe something like "a simple approach to pack sequences greedily without revisiting previous bins like First Fit does."

File: nemo/collections/multimodal/parts/utils.py [hunk 4]
```diff
@@ -320,7 +320,7 @@
 
 
 def create_neva_model_and_processor(cfg):
-    from nemo.collections.multimodal.models.neva.neva_model import MegatronNevaModel
+    from nemo.collections.multimodal.models.multimodal_llm.neva.neva_model import MegatronNevaModel
```
[the remark this answers is not in the document]
Comment (cuichenx) [reply nemo/collections/multimodal/parts/utils.py:323]: could you check this?

**Review by cuichenx — APPROVED** [review]
LGTM!

**Comment by yaoyu-33** [comment]
jenkins

**Comment by yaoyu-33** [comment]
jenkins

**Comment by yaoyu-33** [comment]
jenkins

**Comment by yaoyu-33** [comment]
jenkins

**Review by Victor49152 — APPROVED** [review]

**Review by Victor49152 — APPROVED** [review]

**Pull request merged**